### ISSUE TITLE
Incomplete GC on S3 backend

### ISSUE THREAD

**Issue by mhkarimi1383** [distribution/distribution#4382]
### Description

I have a registry with S3 backend for storage but when I run GC, some directories will remain and List API will return that removed tag/manifest.

### Reproduce

1. Start registry v2 with MinIO as S3 Storage Backend
2. Push some images (preferably with multiple tags)
3. Remove a tag
4. Get list of tags/images

### Expected behavior

Remove everything related to image from S3

### registry version

2.8.3

**Comment by milosgajdos** [comment]
Please don't use `v2` release. Use v3-alpha. v2 releases havent received any update for a very long time and will no longer receive any.  We will be marking it as deprecated once the v3 is out. No ETA.

**Comment by mhkarimi1383** [comment]
@milosgajdos 
There is no ETA for v2 release to be marked as stable?

**Comment by milosgajdos** [comment]
`v2` is marked as stable, but it's not actively maintained. NO ETA on v3, we're hoping to get v3-beta out before summer. Wer'e waiting for one more PR before we can cut the beta.

You can get the v3-alpha or edge from ghcr:
https://github.com/distribution/distribution/pkgs/container/distribution

**Comment by mhkarimi1383** [comment]
@milosgajdos 
Thanks but the link returns `404 not found`

**Comment by milosgajdos** [comment]
yeah I just noticed that...I need to open a ticket with GH.  In the meantime you can pull it from Hub https://hub.docker.com/layers/distribution/distribution/3.0.0-alpha.1/images/sha256-814a7366834827192400baf7b6d2d27e3631a02493570cf9efb697dd75511942?context=explore

**Comment by MalteMagnussen** [comment]
@milosgajdos - How would we go about writing our own Garbage Collection for harbor using v3? 

We have S3 backend for harbor, and are having issues with GC.  
One manifest deletion with v2 can take up to 20 minutes. 

Is it possible to write our own GC with v3? 

Assume I know nothing about distribution, and only discovered today that it was under the hood of harbor. idk how it works.

**Comment by mhkarimi1383** [comment]
> @milosgajdos - How would we go about writing our own Garbage Collection for harbor using v3? 
> 
> We have S3 backend for harbor, and are having issues with GC.  
> One manifest deletion with v2 can take up to 20 minutes. 
> 
> Is it possible to write our own GC with v3? 
> 
> Assume I know nothing about distribution, and only discovered today that it was under the hood of harbor. idk how it works. 

I was able to run GC on the v2 registry with v3 binaries :)

**Comment by MalteMagnussen** [comment]
@mhkarimi1383 - Nice! Did that improve performance for you? :)

**Comment by mhkarimi1383** [comment]
> @mhkarimi1383 - Nice! Did that improve performance for you? :) 

It was way faster

**Comment by mhkarimi1383** [comment]
@milosgajdos 

In the latest version we are still having the same problem

GCs are faster, but we are getting removed things in the list

**Comment by MalteMagnussen** [comment]
Apparently it would be "catastrophic" to use the registry binary for gc on harbor: https://github.com/goharbor/harbor/issues/11606
